From 95f59de8e88f0eb10f1db0a64d99b79a2eeee236 Mon Sep 17 00:00:00 2001
From: Claudiu Beznea <claudiu.beznea@microchip.com>
Date: Fri, 20 Sep 2019 12:26:46 +0300
Subject: [PATCH 258/471] drm/atmel-hlcdc: fix for all !sam9x60

gdx2d was written to work only on sam9x60. Adapt HLCDC code to also
work for non sam9x60 SoCs.

Fixes: 90c3967f2eb1 ("drm/atmel-hlcdc: WIP: gfx2d on 9x60ek hardware")
Tested-by: Ludovic Desroches <ludovic.desroches@microchip.com>
Signed-off-by: Claudiu Beznea <claudiu.beznea@microchip.com>
---
 drivers/gpu/drm/atmel-hlcdc/atmel_hlcdc_dc.c | 35 +++++++++++++++++---
 1 file changed, 31 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/atmel-hlcdc/atmel_hlcdc_dc.c b/drivers/gpu/drm/atmel-hlcdc/atmel_hlcdc_dc.c
index 21221e408aca..e5a4dc1553d3 100644
--- a/drivers/gpu/drm/atmel-hlcdc/atmel_hlcdc_dc.c
+++ b/drivers/gpu/drm/atmel-hlcdc/atmel_hlcdc_dc.c
@@ -956,6 +956,8 @@ static int atmel_hlcdc_dc_drm_probe(struct platform_device *pdev)
 {
 	struct component_match *match = NULL;
 	struct device_node *core_node;
+	struct drm_device *ddev;
+	int ret;
 
 	for_each_compatible_node(core_node, NULL, "microchip,sam9x60-gfx2d") {
 		if (!of_device_is_available(core_node))
@@ -965,11 +967,36 @@ static int atmel_hlcdc_dc_drm_probe(struct platform_device *pdev)
 				    compare_of, core_node);
 	}
 
-	if (!match)
-		return -ENODEV;
+	if (match) {
+		return component_master_add_with_match(&pdev->dev,
+						       &atmel_hlcdc_dc_master_ops,
+						       match);
+	}
+
+	/* Fall through old probe routine */
+	ddev = drm_dev_alloc(&atmel_hlcdc_dc_driver, &pdev->dev);
+	if (IS_ERR(ddev))
+		return PTR_ERR(ddev);
+
+	ret = atmel_hlcdc_dc_load(ddev);
+	if (ret)
+		goto err_put;
+
+	ret = drm_dev_register(ddev, 0);
+	if (ret)
+		goto err_unload;
+
+	drm_fbdev_generic_setup(ddev, 24);
+
+	return 0;
 
-	return component_master_add_with_match(&pdev->dev,
-					       &atmel_hlcdc_dc_master_ops, match);
+err_unload:
+	atmel_hlcdc_dc_unload(ddev);
+
+err_put:
+	drm_dev_put(ddev);
+
+	return ret;
 }
 
 static int atmel_hlcdc_dc_drm_remove(struct platform_device *pdev)
-- 
2.34.1

